packages/devtools/src/schema/compiler.ts

```diff
@@ -0,0 +1,328 @@
+import type { SetRequired } from 'type-fest'
+
+import { DOC_ID_FIELD } from '../constants.js'
+import type { AnySchema, ScalarSchema } from '../types.js'
+
+import { parseSchema } from './parser.js'
+import { isCommonScalar } from './scalars.js'
+import type {
+  AbstractCompositeDefinition,
+  AbstractModelDefinition,
+  EnumFieldDefinition,
+  ListFieldDefinition,
+  ObjectDefinition,
+  ObjectReferenceFieldDefinition,
+  ParsedModelDefinition,
+  ScalarFieldDefinition,
+  SchemaDefinition,
+} from './types.js'
+
+type CompileObject = SetRequired<JSONSchema.Object, 'properties'>
+
+type JSONSchemaReference = { $ref: string }
+
+type ReferencedSchema = ScalarSchema | JSONSchema.Object
+
+type SchemaWithRefs<T extends AnySchema = AnySchema> = { schema: T; refs: Array<string> }
+
+export function createReference(name: string): JSONSchemaReference {
+  return { $ref: `#/$defs/${name}` }
+}
+
+export function extractReference(ref: JSONSchemaReference): string {
+  const name = ref.$ref.split('#/$defs/')[1]
+  if (name == null || name.length === 0) {
+    throw new Error(`Could not extract name from reference: ${ref.$ref}`)
+  }
+  return name
+}
+
+export class SchemaCompiler {
+  #def: AbstractCompositeDefinition = {
+    models: {},
+    commonEmbeds: [],
+  }
+  #refs: Record<string, SchemaWithRefs<ReferencedSchema>> = {}
+  #src: SchemaDefinition
+
+  constructor(source: SchemaDefinition) {
+    this.#src = source
+  }
+
+  _getReference(name: string): SchemaWithRefs<ReferencedSchema> {
+    const ref = this.#refs[name]
+    if (ref == null) {
+      throw new Error(`Reference ${name} does not exist`)
+    }
+    return ref
+  }
+
+  _extractDefinitions(name: string): Record<string, ReferencedSchema> {
+    const unref = this._getReference(name)
+    const defs: Record<string, ReferencedSchema> = { [name]: unref.schema }
+    for (const subrefName of unref.refs) {
+      Object.assign(defs, this._extractDefinitions(subrefName))
+    }
+    return defs
+  }
+
+  compile(): AbstractCompositeDefinition {
+    // Ensure enums are tracked in common embeds
+    for (const name of Object.keys(this.#src.enums)) {
+      this.#def.commonEmbeds.push(name)
+    }
+    // Only compile embedded objects in first pass so they can be added to models in second pass
+    for (const [name, definition] of Object.entries(this.#src.objects)) {
+      if (this.#src.models[name] == null) {
+        this._compileEmbedObject(name, definition)
+        this.#def.commonEmbeds.push(name)
+      }
+    }
+    // Compile models
+    for (const [name, definition] of Object.entries(this.#src.models)) {
+      const object = this.#src.objects[name]
+      if (object == null) {
+        throw new Error(`Missing object definition for model: ${name}`)
+      }
+      // Compile object schema with embedded references for model
+      this.#def.models[name] = this._compileModel(name, definition, object)
+    }
+    return this.#def
```

Comment:
Why do we need this `#def` property which is being mutated in `compile()`?

It feels to me like this may be quite error-prone, as other methods than `compile()` in the `SchemaCompiler` class can reference `#def`, but they have no way to know, if the value of `#def` is in a consistent state.

Reply:
That's the whole point of the compiler: to create the `AbstractCompositeDefinition` returned here. The `compile()` method just needs to be called once.

Reply:
Yes, but why do we need to store this mutable `#def` property rather than just create the definition in the `compile()` method and return it?